service/controller/v15/drainer_resource_set.go

```diff
@@ -33,21 +35,48 @@ type DrainerResourceSetConfig struct {
 func NewDrainerResourceSet(config DrainerResourceSetConfig) (*controller.ResourceSet, error) {
 	var err error
 
-	var lifecycleResource controller.Resource
+	var drainerResource controller.Resource
 	{
-		c := lifecycle.ResourceConfig{
+		c := drainer.ResourceConfig{
 			G8sClient: config.G8sClient,
 			Logger:    config.Logger,
 		}
 
-		lifecycleResource, err = lifecycle.NewResource(c)
+		drainerResource, err = drainer.NewResource(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var drainFinisherResource controller.Resource
+	{
+		c := drainfinisher.ResourceConfig{
+			G8sClient: config.G8sClient,
+			Logger:    config.Logger,
+		}
+
+		drainFinisherResource, err = drainfinisher.NewResource(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var workerASGNameResource controller.Resource
+	{
+		c := workerasgname.ResourceConfig{
+			Logger: config.Logger,
+		}
+
+		workerASGNameResource, err = workerasgname.NewResource(c)
 		if err != nil {
 			return nil, microerror.Mask(err)
 		}
 	}
 
 	resources := []controller.Resource{
-		lifecycleResource,
+		workerASGNameResource,
+		drainerResource,
+		drainFinisherResource,
 	}
```

Comment:
I split `lifecycle` resource into three separate ones:

- `workerasgname` - retrieves worker ASG name from CF stack when it is ready. Otherwise it cancels reconciliation early. This reduces number of logs and API calls.
- `drainer` - creates DrainerConfigs for ASG instances in terminating state.
- `drainFinisher` - completes ASG lifecycle hooks for drained nodes, and then deletes drained DrainerConfigs.

TODO I should add that to godocs :)

Reply:
Done.